examples/raytracing/example_raytracing.cpp

```diff
@@ -0,0 +1,232 @@
+
+    //  for normalized direction vector a = 1.0
+    auto dir = ray.direction();
+
+    float b = 2.0 * dotproduct(dir, L);
+    float c = dotproduct(L, L) - r * r;
+
+    return solveQuadratic(b, c);
+  }
+
+  template <typename Predicate>
+  KOKKOS_FUNCTION void operator()(Predicate const &predicate,
+                                  int primitive_index) const
+  {
+    auto const &ray = ArborX::getGeometry(predicate);
+    auto const &sphere = spheres(primitive_index);
+    float length = overlap_distance(ray, sphere);
+    int i = getData(predicate);
+    accumulator(i) += length;
+
+    //#ifndef __SYCL_DEVICE_ONLY__
+    //    printf("ray %d hit sphere %d with the overlap %f. \n", i,
+    //    primitive_index, length);
+    //#endif
+  }
+};
+
+int main(int argc, char *argv[])
+{
+  using ExecutionSpace = Kokkos::DefaultExecutionSpace;
+  using MemorySpace = ExecutionSpace::memory_space;
+
+  Kokkos::ScopeGuard guard(argc, argv);
+
+  namespace bpo = boost::program_options;
+
+  int num_spheres;
+  int num_rays;
+  float L;
+
+  bpo::options_description desc("options");
+  // clang-format off
+  desc.add_options()
+    ("num_spheres", bpo::value<int>(&num_spheres), "number of spheres")
+    ("num_rays", bpo::value<int>(&num_rays), "number of rays")
+    ("L", bpo::value<float>(&L), "size of the domain")
+    ;
+
+  bpo::variables_map vm;
+  bpo::store(bpo::command_line_parser(argc, argv).options(desc).run(), vm);
+  bpo::notify(vm);
+
+  std::uniform_real_distribution<float> uniform{0.0, 1.0};
+  std::default_random_engine gen;
+  auto rd_uniform = [&]() { return uniform(gen); };
+
+  // radius
+  const float mu_R = 1.0;
+  const float sigma_R = mu_R / 3.0;
+
+  std::normal_distribution<> normal{mu_R, sigma_R};
+  auto rd_normal = [&]() { return std::max(normal(gen), 0.0); };
+
+  Kokkos::View<ArborX::Sphere *, MemorySpace> spheres_view("spheres_view", num_spheres);
+  auto spheres_host = Kokkos::create_mirror_view(spheres_view);
+  for (int i = 0; i < num_spheres; ++i)
+  {
+    spheres_host[i] = {{rd_uniform(), rd_uniform(), rd_uniform()}, rd_normal()};
+  }
+  Kokkos::deep_copy(spheres_view, spheres_host);
+  Primitives<MemorySpace> spheres{spheres_view};
+  ArborX::BVH<MemorySpace> bvh{ExecutionSpace{}, spheres};
+
+  Kokkos::View<ArborX::Ray *, MemorySpace> rays_view("rays_view", num_rays);
+  auto rays_host = Kokkos::create_mirror_view(rays_view);
+
+  for (int i = 0; i < num_rays; ++i)
+  {
+    float x = rd_uniform() * L;
+    float y = rd_uniform() * L;
+    float z = 0.0;
+
+    float sinazimuth = std::sin(rd_uniform() * 2.0 * M_PI);
+    float cosazimuth = std::sqrt(1 - sinazimuth * sinazimuth);
+    float cospolar = 1.0 - rd_uniform();
+    float sinpolar = std::sqrt(1.0 - cospolar * cospolar);
+
+    float dirx = sinpolar * cosazimuth;
+    float diry = sinpolar * sinazimuth;
+    float dirz = cospolar;
+
+    rays_host[i] = {{x, y, z}, {dirx, diry, dirz}};
```

Comment:
No idea what's happening here, please add a comment describing what and how is being constructed.